Tensile/KernelWriterAssembly.py

```diff
@@ -3856,6 +3895,52 @@ def graWorkGroup(self, kernel, isPap):
+        # if stmp+1 < stmp+3, continue DP (add dpShift)
+        kStr += inst("s_cmp_lt_u32", sgpr(stmp+1), sgpr(stmp+3), "Check if still in DP section")
+        kStr += inst("s_cbranch_scc1 %s" % skUpdateDone, "Done update")
+        # if StreamKIter >= stmp+3, continue SK (add skShift?)
+        kStr += inst("s_mov_b32", sgpr(stmp+1), sgpr(stmp+2), "SK iterations shift")
+        kStr += inst("s_cmp_le_u32", sgpr(stmp+3), sgpr("StreamKIter"), "Check if continuing in SK section")
+        kStr += inst("s_cbranch_scc1 %s" % skUpdateDone, "Done update")
+        # if stmp+1 > stmp+3 and StreamKIter < stmp+3, switch from DP to SK (add dpShift)
+        # iter count after all extra iters have been distributed
+        kStr += inst("s_mul_i32", sgpr("StreamKIter"), sgpr("StreamKIdx"), sgpr("SKItersPerWG"), "StreamK starting iteration (case: after extra iters)")
+        kStr += inst("s_add_u32", sgpr("StreamKIter"), sgpr("StreamKIter"), sgpr("skExtraIters"), "Add extra iters")
+        kStr += inst("s_add_u32", sgpr("StreamKIterEnd"), sgpr("StreamKIter"), sgpr("SKItersPerWG"), "StreamK ending iteration (case: after extra iters)")
+        # iter count before all extra iters have been distributed
+        # stmp+1 = SKItersPerWG + 1 extra iteration
+        sIter = self.sgprPool.checkOut(2, "SKIter", preventOverflow=0)
+        kStr += inst("s_add_u32", sgpr(sIter+1), sgpr("SKItersPerWG"), 1, "Spread out extra iterations")
+        kStr += inst("s_mul_i32", sgpr(sIter), sgpr("StreamKIdx"), sgpr(sIter+1), "StreamK starting iteration (case: before extra iters)")
+        kStr += inst("s_add_u32", sgpr(sIter+1), sgpr(sIter), sgpr(sIter+1), "StreamK ending iteration (case: before extra iters)")
+        # select correct start/end iteration index
+        kStr += inst("s_cmp_lt_u32", sgpr("StreamKIdx"), sgpr("skExtraIters"), "Check if lane gets an extra iteration")
+        kStr += inst("s_cselect_b32", sgpr("StreamKIter"), sgpr(sIter), sgpr("StreamKIter"), "Set start iter")
+        kStr += inst("s_cselect_b32", sgpr("StreamKIterEnd"), sgpr(sIter+1), sgpr("StreamKIterEnd"), "Set end iter")
+        self.sgprPool.checkIn(sIter)
+        kStr += inst("s_add_u32", sgpr(stmp+1), sgpr("StreamKIter"), sgpr(stmp+3), "Offset to start of SK section")
+        kStr += inst("s_add_u32", sgpr("StreamKIterEnd"), sgpr("StreamKIterEnd"), sgpr(stmp+3), "Offset to start of SK section")
+        # clamp to end of iterations
+        # TODO maybe remove clamp, since extra iters code should guarantee total iterations match
+        kStr += inst("s_min_u32", sgpr("StreamKIterEnd"), sgpr("StreamKIterEnd"), sgpr("TotalIters"), "Cap ending iter at total SK iters")
+        # check if this WG has no work to do
+        # TODO Shouldn't need this check!
+        kStr += inst("s_cmp_lt_u32", sgpr("StreamKIter"), sgpr("TotalIters"), "Make sure there's work to do")
+        kStr += self.longBranchScc0("label_%04u" % (self.getLabelNum("KernelEnd")), positiveOnly=True)
```

Comment:
Long jump code uses 3 sgpr tmp and increases sgpr usage of this kernel.
Is that possible to release stmp before long jump generation and re-allocate same stmp after long jump code is generated?
This long jump goes to kernel end. No need to keep any values.

Reply:
Tried a couple ways to change this, but wasn't happy with either change, will send more details.

Reply:
OK. In that case, we need to add an optional parameter in long jump function to optionally pass tmp sgpr registers to avoid allocating extra sgpr registers.
But, this is an optional.
I am OK to merge the current change for now.

Reply:
Just thought the same thing :D going to post the change shortly

Reply:
Posted the new change